File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-8.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # substring = consecutive chars
        # use a set to keep track of what chars are in the window
        '''
        # just return the length
        # use a window to keep track of the longest substring
        for loop for the right pointer
        while loop inside for the left pointer (increment the left until the window is valid again (no char duplicates))


        abba
          ^^
        seen = { a:0, b: 2,  }
        length = 2
        maxLength = 1
        '''
        left = 0
        seen = {}
        maxLen = length = 0
        for i in range(0, len(s)):
            length += 1
 
            if s[i] in seen:
                consider = seen[s[i]] + 1
                left = max(consider, left)
                length = i - left + 1

            maxLen = max(maxLen, length)

            seen[s[i]] = i
        return maxLen

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-8.js
class Solution {
    /**
     * @param {number[]} prices
     * @return {number}
     */
    maxProfit(prices) {
        let maxProfit = 0
        let left = 0
        for (let right = 0; right < prices.length; right++){
            let total = prices[right] - prices[left]
            if (prices[left] > prices[right]){
                left = right
            }
            maxProfit = Math.max(total, maxProfit)

        }
        return maxProfit
    }
}
